kyaml/kio/byteio_reader.go

```diff
@@ -282,3 +299,46 @@ func (r *ByteReader) decode(index int, decoder *yaml.Decoder) (*yaml.RNode, erro
 	}
 	return yaml.NewRNode(node), nil
 }
+
+// seqIndentAnno derives the sequence indentation annotation value for the resource,
+// originalYAML is the input yaml string and node is the decoded equivalent of originalYAML,
+// the annotation value is decided by deriving the existing sequence indentation of resource
+func seqIndentAnno(node *yaml.Node, originalYAML string) string {
+	rNode := &yaml.RNode{}
+	rNode.SetYNode(node)
+	// step 1: derive the sequence indentation of the node
+	anno := rNode.GetAnnotations()
+	if anno[kioutil.SeqIndentAnnotation] != "" {
+		// the annotation already exists, so don't change it
+		return ""
+	}
+
+	// marshal the node with 2 space sequence indentation and calculate the diff
+	out, err := yaml.MarshalWithOptions(rNode.Document(), &yaml.EncoderOptions{
+		MapIndent: yaml.DefaultMapIndent,
+		SeqIndent: yaml.WideSeqIndent,
+	})
+	if err != nil {
+		return ""
+	}
+	twoSpaceIndentDiff := copyutil.PrettyFileDiff(string(out), originalYAML)
+
+	// marshal the node with 0 space sequence indentation and calculate the diff
+	out, err = yaml.MarshalWithOptions(rNode.Document(), &yaml.EncoderOptions{
```

Comment:
This feature requiring two extra marshallings of every resource really makes me wish we had perf benchmarks to see if that is impactful.

Reply:
This only happens if callers opt in for preserving indent.